File: src/main/de/rub/bi/inf/openbimrl/functions/ifc/GetPropertyAsString.java
package de.rub.bi.inf.openbimrl.functions.ifc;

import de.rub.bi.inf.openbimrl.NodeProxy;
import de.rub.bi.inf.openbimrl.engine.ifc.IIFCModel;
import de.rub.bi.inf.openbimrl.functions.AbstractFunction;
import de.rub.bi.inf.openbimrl.helper.IfcUtilities;

import java.util.ArrayList;
import java.util.Collection;

/**
 * Retrieves the property value of a specific entity as String value.
 *
 * @author Marcel Stepien
 */
public class GetPropertyAsString extends AbstractFunction {

    public GetPropertyAsString(NodeProxy nodeProxy) {
        super(nodeProxy);
    }

    @Override
    public void execute(IIFCModel ifcModel) {

   /*     Collection<Object> objects = new ArrayList();
        Object input0 = getInput(0);
        if (input0 instanceof Collection<?>) {
            objects = (Collection<Object>) input0;
        } else {
            objects.add(input0);
        }

        String setName = getInput(1);
        String propertyName = getInput(2);

        ArrayList<Object> resultValues = new ArrayList<>();

        for (Object o : objects) {
            IfcObjectDefinition.Ifc4 ifcObjectDefinition = (IfcObjectDefinition.Ifc4) o;
            IfcSimpleProperty.Ifc4 simpleProperty = IfcUtilities.Ifc4.getElementProperty(
                    ifcObjectDefinition, setName, propertyName);

            if (simpleProperty instanceof IfcPropertySingleValue.Ifc4) {
                IfcPropertySingleValue.Ifc4 singleValue =
                        (IfcPropertySingleValue.Ifc4) simpleProperty;
                resultValues.add(singleValue.getNominalValue().toString());
            } else if (simpleProperty instanceof IfcPropertyEnumeratedValue.Ifc4) {
                IfcPropertyEnumeratedValue.Ifc4 enumeratedValue =
                        (IfcPropertyEnumeratedValue.Ifc4) simpleProperty;
                resultValues.add(enumeratedValue.getEnumerationValues().get(0).toString());
            } else {

                //FIXME other possibilities
                resultValues.add(null);

            }

        }

        if (resultValues.size() == 1) {
            setResult(0, resultValues.get(0));
        } else {
            setResult(0, resultValues);
        }*/

    }

}

File: src/main/jvm/de/rub/bi/inf/openbimrl/functions/ifc/GetPropertyAsString.kt
package de.rub.bi.inf.openbimrl.functions.ifc

import de.rub.bi.inf.nativelib.IfcPointer
import de.rub.bi.inf.openbimrl.NodeProxy
import de.rub.bi.inf.openbimrl.functions.AbstractFunction
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionInput
import de.rub.bi.inf.openbimrl.functions.annotations.FunctionOutput
import de.rub.bi.inf.openbimrl.functions.annotations.OpenBIMRLFunction

@OpenBIMRLFunction(description = "Retrieves the property value of a specific entity as String value.")
class GetPropertyAsString(nodeProxy: NodeProxy) : AbstractFunction(nodeProxy) {

    @FunctionInput(0, name = "IfcElement", collectionType = IfcPointer::class)
    lateinit var ifcElements: Collection<IfcPointer>

    @FunctionInput(1, name = "PropertySet_Name")
    lateinit var propertySetName: String

    @FunctionInput(2, name = "Property_Name")
    lateinit var propertyName: String

    @FunctionOutput(0)
    var value: Any? = null

    override fun execute() {
        val resolved = ifcElements.filter { element ->
            element.properties.containsKey(propertySetName) &&
                element.properties[propertySetName]!!.containsKey(propertyName)
        }.map { element ->
            element.properties[propertySetName]!![propertyName]!!
        }

        value = if (resolved.size == 1) resolved[0] else resolved
    }
}
